pkg/mutation/mutate.go

```diff
@@ -116,3 +132,225 @@ func UpdateMutatedContentWithComments(yamlContent string, comments []config.Muta
 	}
 	return fileContent
 }
+
+func findNodes(node *yaml.Node, selectors []string) ([]*yaml.Node, error) {
+	var nodes []*yaml.Node
+	currentSelector := selectors[0]
+	// array[N] or array[*] selectors.
+	if i := strings.LastIndex(currentSelector, "["); i > 0 && strings.HasSuffix(currentSelector, "]") {
+		arrayIndex := currentSelector[i+1 : len(currentSelector)-1]
+		currentSelector = currentSelector[:i]
+
+		index, err := strconv.Atoi(arrayIndex)
+		if err != nil {
+			if arrayIndex == "*" {
+				index = -1
+			} else {
+				return nil, errors.Wrapf(err, "can't parse array index from %v[%v]", currentSelector, arrayIndex)
+			}
+		} else if index < 0 {
+			return nil, errors.Wrapf(err, "array index can't be negative %v[%v]", currentSelector, arrayIndex)
+		}
+
+		// Go into array node(s).
+		arrayNodes, err := findNodes(node, []string{currentSelector})
+		if err != nil {
+			return nil, errors.Errorf("can't find %v", currentSelector)
+		}
+		for _, arrayNode := range arrayNodes {
+			if arrayNode.Kind != yaml.SequenceNode {
+				return nil, errors.Errorf("%v is not an array", currentSelector)
+			}
+			if index >= len(arrayNode.Content) {
+				return nil, errors.Errorf("%v array doesn't have index %v", currentSelector, index)
+			}
+
+			var visitArrayNodes []*yaml.Node
+			if index >= 0 { // array[N]
+				visitArrayNodes = []*yaml.Node{arrayNode.Content[index]}
+			} else { // array[*]
+				visitArrayNodes = arrayNode.Content
+			}
+
+			for i, node := range visitArrayNodes {
+				if len(selectors) == 1 {
+					// Last selector, use this as final node.
+					nodes = append(nodes, node)
+				} else {
+					// Go deeper into a specific array.
+					deeperNodes, err := findNodes(node, selectors[1:])
+					if err != nil {
+						return nil, errors.Wrapf(err, "failed to go deeper into %v[%v]", currentSelector, i)
+					}
+					nodes = append(nodes, deeperNodes...)
+				}
+			}
+		}
+		return nodes, nil
+	}
+	switch node.Kind {
+	case yaml.MappingNode:
+		for i := 0; i < len(node.Content); i += 2 {
+			// Does the current key match the selector?
+			if node.Content[i].Value == currentSelector {
+				// Found last key, return its value.
+				isLastSelector := len(selectors) == 1
+				if !isLastSelector {
+					// Match the rest of the selector path, ie. go deeper
+					// in to the value node.
+					return findNodes(node.Content[i+1], selectors[1:])
+				}
+				return []*yaml.Node{node.Content[i+1]}, nil
+			}
+		}
+	case yaml.ScalarNode:
+		// Overwrite any existing nodes.
+		node.Kind = yaml.MappingNode
+		node.Tag = "!!map"
+		node.Value = ""
+	case yaml.SequenceNode:
+		return nil, errors.Errorf("parent node is array, use [*] or [0]..[%v] instead of .%v to access its item(s) first", len(node.Content)-1, currentSelector)
+
+	default:
+		return nil, errors.Errorf("parent node is of unknown kind %v", node.Kind)
+	}
+	// Create the rest of the selector path.
+	for _, selector := range selectors {
```

Comment:
It's interesting that `findNodes` _also_ modifies the nodes.

I wonder if we should break out an `ensurePathExists`, or at least rename `findNodes` to something like `createPathAndFindNodes`